LoRa radio capsule and nRF52 board i/f.

nikulshr opened the pull request:
### Pull Request Overview

Adds a LoRa interface for mirco-controlling a LoRa modem (SX127X) using SPI. So far, the PHY layer has been implemented. Note: Changes to the nRF52 board are mainly to add the underlying SPI mux and driver component.

### Testing Strategy

Tested on MCCI Catena 4610 (SX1276) using Nordic's nRF52840dk by checking for SPI communication over a logic analyzer. TBD: Check for actual OTA packets.

### TODO or Help Wanted

Needs some review of the existing SPI (enable certain blocks currently commented, cleaner code, bugs), general Tock interfaces and possibly LoRa functions.

### Documentation Updated

- Added README under capsules/src/lora

### Formatting
- Ran `make formatall`

brghena commented:
Welcome @nikulshr! I'm really excited to see some LoRa radio support for Tock.

For pull requests, we've got a [Github PR template](https://github.com/tock/tock/blob/master/.github/PULL_REQUEST_TEMPLATE.md) which you should be filling out. It should have appeared automatically when you made the PR. Please edit the description of this and fill it out so we know the status for this PR.

bradjc commented on capsules/src/lora/radio.rs:
```diff
@@ -0,0 +1,761 @@
+      //    return ReturnCode::EBUSY;
+      //}
+      let wbuf = self.spi_tx.take().unwrap();
+      let rbuf = self.spi_rx.take().unwrap();
```
I think this can be:
```rust
self.spi_tx.take().map(|wbuf| {
    self.spi_rx.take().map(|rbuf| { ... }
)});
```

nikulshr replied:
Any advantage of this particular syntax? I find the current one much cleaner and also existing elsewhere in Tock.

lschuermann replied:
Using `.map` automatically puts the buffers back into the cell after the closure execution. Using `.take()` will probably make you forget that at some point. `.take()` is used elsewhere when the buffers are passed to components returning them in a callback, as the references in `.map()` will only be valid for the closure and therefore can't be moved out of it.

bradjc commented on chips/nrf52/src/spi.rs:
```diff
@@ -382,7 +383,9 @@ impl hil::spi::SpiMaster for SPIM {
         self.busy.set(true);
+      debug_gpio!(1, toggle);
         self.registers.tasks_start.write(TASK::TASK::SET);
+      debug_gpio!(1, toggle);
```
remove

nikulshr replied:
Done

bradjc commented on capsules/src/spi.rs:
```diff
@@ -121,6 +122,7 @@ impl<S: SpiMasterDevice> Driver for Spi<'a, S> {
             }
             // Pass in a write buffer to transmit bytes from.
             1 => {
+
```
undo add newlines

nikulshr replied:
Done

nikulshr commented on boards/nordic/nrf52840dk/src/main.rs:
```diff
@@ -88,14 +88,18 @@ const UART_TXD: Pin = Pin::P0_06;
-const SPI_CLK: Pin = Pin::P0_19;
+const SPI_MOSI: Pin = Pin::P0_29;
+const SPI_MISO: Pin = Pin::P0_30;
+const SPI_CLK: Pin = Pin::P0_26;
```
In my case, I couldn't detect SPI on these pins whereas when I matched them to the nRF52 DK default, it worked! Since I had to keep the rest of the changes, I have just reverted these lines for now with that comment.

nikulshr commented:
Thanks all for comments so far! I have updated description as suggested by @brghena and also added comments like @bradjc said. P.S. Still new to Git; hope everything is in place

brghena commented:
Thanks for the updates @nikulshr! Two notes:

1. You seemed to have picked up extraneous commits into this PR. It looks like they were other PRs that were merged into Tock master recently. Perhaps something went wrong in a rebase? Please figure out what's going on and update the PR. It's pretty hard to see your changes out of the 19 files it thinks are changed.

2. The Formatting is definitely applicable to this PR. Be sure to add it back to the PR description and run `make format` from the home directory of Tock to apply our Rust formatting standards.

nikulshr commented:
> Thanks for the updates @nikulshr! Two notes:
> 
> 1. You seemed to have picked up extraneous commits into this PR. It looks like they were other PRs that were merged into Tock master recently. Perhaps something went wrong in a rebase? Please figure out what's going on and update the PR. It's pretty hard to see your changes out of the 19 files it thinks are changed.
> 2. The Formatting is definitely applicable to this PR. Be sure to add it back to the PR description and run `make format` from the home directory of Tock to apply our Rust formatting standards.

1. Yeah, apologize for that. I've taken care of it (took me an hour *sigh*). For anybody who might benefit: just `git reset` to YOUR last commit and unstage any extraneous local changes.
2. I wasn't aware of this check; how can I know more? I ran it and got this:
`fatal: There is nothing to exclude from by :(exclude) patterns.
Perhaps you forgot to add either ':/' or '.' ?`

hudson-ayers commented:
> 2. I wasn't aware of this check; how can I know more? I ran it and got this:
>    `fatal: There is nothing to exclude from by :(exclude) patterns. Perhaps you forgot to add either ':/' or '.' ?`

`make format` basically runs `rustfmt`, the standard tool for formatting Rust code to meet style guidelines, on all of the rust files in Tock. (It also does a couple other small things like complain if you attempt to use wildcard imports).
I cloned your current branch and ran `make format` and it ran successfully. I looked into that particular error a bit and it looks like the issue is your are probably using a git version older than 2.13, so the syntax for exclusion was stricter (the relevant commit: https://github.com/git/git/commit/859b7f1d0e742493d2a9396794cd9040213ad846). The easiest fix is probably to update git to a newer version.

nikulshr commented:
- So I ran `make format all`. I still get CI failure.
- More suggestions to improve code are welcome :-)

brghena commented:
You can click "Details" on the CI check to see what failed. You can also run `make ci` locally to see what's going on and whether issues are solved without pushing commits to Github.

nikulshr commented:
I'm getting a lot of errors with unused variables/functions. I know it's bad practice, but I want to keep them for future iterations. Is there a way to bypass this? What's the best way to do things here?

bradjc commented:
You can keep them in a separate branch, but we tend to avoid unused things in the tock repo.

bradjc commented:
Proposed roadmap to merge this:
- Remove the changes to the nRF boards since by default those boards won't have a LoRa chip attached.
- Either expand or remove the `Driver` implementation. I think that was good for testing, but it's a little too narrow to include here.
- Rename `lora/radio.rs` to something like `lora/loraspi.rs`, and update the structs to match. The driver is super useful, but it isn't the only way to write a driver for a lora radio, so making it a little more clear makes it easier for the next person to write a lora driver.

Otherwise, I think this looks great, and is a really nice step towards LoRa support.

nikulshr commented:
Two things:
- @bradjc can you clarify the 3rd point on your roadmap? I went with `radio.rs` as it does have radio functionality and not just SPI (used by it).
- the CI is waiting on several checks, not sure I understand why and if they're necessary over `deploy-preview`

ppannuto commented:
I suspect CI might have been confused since the configuration changed in Tock master since this branch was made.

It was also a bit messy to merge after the nrf component stuff went through.

I added a commit that merges Tock `master` in this branch (you should be able to `git pull` locally to get the update). I believe I did everything correctly.

ppannuto commented:
I spent a little bit more time on this this evening. As this driver is designed to talk over a SPI bus to a LoRa module, it shouldn't really be just an nrf component. I refactored the component design to eliminate any specific references to the nrf and moved it to the generic components.

Along the way, I figured out the confusion around the `SpiMasterDevice` trait and was able to remove that trait bound from the structures, which simplified them a bit. [This is the key change](https://github.com/tock/tock/pull/1760/commits/3aaecfcc213832a7c884ec0d256d1b3d8b909818#diff-13fe34f36d1459dba28e9b615521021cR174). Instead of storing a reference to an object whose type is known at compile-time, the `dyn` keyword lets it be a runtime decision. As #1761 notes, there is non-zero overhead to this, but it's also much simpler and it's in line with what the rest of the kernel is doing, so I think it makes sense to do that for now.

hudson-ayers commented:
Merging of this is still blocked on removing the changes in `boards/nordic`, but the recently pushed updates look good. Also, once this is ready, can you add doc comments for the lora HIL?

bradjc commented on boards/nordic/nrf52840_dongle/src/main.rs:
```diff
@@ -292,9 +292,9 @@ pub unsafe fn reset_handler() {
-        button,
         ble_radio,
         ieee802154_radio,
+        button,
```
undo

bradjc commented on capsules/src/lora/radio.rs:
```diff
@@ -0,0 +1,541 @@
+// For details on LoRa parameters, see Semtech SX1276/77/78/79 Datasheet.
```
Can you rename to `sx127x.rs`?

bradjc commented on capsules/src/lora/radio.rs:
```diff
@@ -0,0 +1,541 @@
+const MAX_PKT_LENGTH: u8 = 255;
+
+// The modem
+pub struct Radio<'a> {
```
Same here
```suggestion
pub struct SX127X<'a> {
```

bradjc commented:
Timed out, moved to #2344 to keep track of. Please re-open when it is updated.
